### ISSUE TITLE
Extracting rationales

### ISSUE THREAD

**Issue by Sahandfer** [behavioral-data/Empathy-Mental-Health#5]
Hello, wish you a happy new year and congratulations on your great work.
I wanted to ask whether it's possible to use the model to extract rationales from the test data or any input with a similar structure to it ( id | seeker post| helper response) since at the moment I can only reproduce the three empathy scores.

Thank you

**Comment by Astromis** [comment]
Hi, @Sahandfer .

Seems that authors don't look here, so I'll try to answer.

If you look at the 120th line of the empathy_classifier.py file, which is the output one of the classifiers, you will see this:
https://github.com/behavioral-data/Empathy-Mental-Health/blob/92f13509af90d0725ef8d8f4ff52fdaab5d3758e/src/empathy_classifier.py#L120

As you can see, there is some logits_rationale_ER, that is never used below but it might be what you are looking for. According to the paper, 
![firefox_ICfwddzU4U](https://user-images.githubusercontent.com/30126032/110913650-79c6ce00-8326-11eb-99fb-aa6d7b42faf0.png)

and if you trace the origins of logits_rationale_ER, you'll see this

https://github.com/behavioral-data/Empathy-Mental-Health/blob/92f13509af90d0725ef8d8f4ff52fdaab5d3758e/src/models/models.py#L215

which is seems to be what being said in the paper.

So, all you need to do is to make sure, that this output is really a mask which is about said in paper and somehow represent the values of this mask according to input token sequenes.

**Comment by Sahandfer** [comment]
Yes, I had seen those lines of code yet failed to get anything useful from the logits. I guess I'll try to see if I made any mistakes. Thank you for your response. I appreciate it!

**Comment by Astromis** [comment]
@Sahandfer 

If you will try again, please, share the sample of logits_rationales ouput.

**Comment by Sahandfer** [comment]
@Astromis Sure! here you go:

Post: I lost my father today. He was my best friend.
Response: Sorry to hear that. That must be really hard.

For the ER classifier, I get a score of 2 and the following logits_rationale:
[[[0.32529351115226746, -0.005434087943285704],
  [0.2360060214996338, -0.004176668357104063],
  [0.23904669284820557, -0.06472361832857132],
  [0.2321399748325348, -0.06611941009759903],
  [0.23249341547489166, -0.07575947791337967],
  [0.259364515542984, -0.07348399609327316],
  [0.24772581458091736, -0.03236930072307587],
  [0.27259010076522827, -0.027884982526302338],
  [0.24484774470329285, -0.03548089414834976],
  [0.24770760536193848, -0.041797906160354614],
  [0.26260143518447876, -0.06285354495048523],
  [0.3051230311393738, 0.2664923667907715],
  [0.2779783606529236, -0.1588348001241684],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293710112571716, -0.010398931801319122],
  [0.32293710112571716, -0.010398931801319122],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293635606765747, -0.010399304330348969],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293641567230225, -0.010399214923381805],
  [0.32293710112571716, -0.010398931801319122],
  [0.32293710112571716, -0.010398931801319122]]]

**Comment by Astromis** [comment]
@Sahandfer 

1) As you can see, the output is not the valid probability distribution. To fix this, you just need to apply softmax function to the logits

`F.softmax(logits_rationale, dim=1)`

2) You can see, that starting from 13 instance, there are same values [0.32293710112571716, -0.010398931801319122]. It's because the length of sequence is fixed up to some number, you can find it in the code or just view the shape. If sequence is shorter than that value, the sequence is padded with the some fixed token e. g. [PAD].
If you count "tokens" In the response it will be 11 and if you count pairs in the logits that doesn't equal [0.32293710112571716, -0.010398931801319122], it will be 13. Why? Because, as minimum, bert bases models added special tokens like [CLS] and it might be that some word was divided by encoder. 

All in all, you need to find encoder instance, the input sequence that was encoded and decode it back. After that, you can cut all pad-tokens and take slice of ratiobalies with valid tokens.

**Comment by Sahandfer** [comment]
@Astromis Thank you for your response. I'm afraid there has been a misunderstanding so let me clarify.
1. I have applied softmax to the logits; however, as you asked me to share the logits, I provided the logits prior to the softmax function.
2. Upon applying softmax on the logits and calculating the argmax, I get the following rationales:
 [[0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], which suggests that most of the input sequence is the rationale for this er score. However, upon adding more sentences like 'Were you two close?" to the response, the additional parts would also be considered as a rationale for this score. Hence, I was unable to rather accurately find the rationale for these scores.

P.S: for the following response "Sorry to hear that. That must be really hard. Were you two close? I can't imagine losing my dad.", the output rationale is as follows:
[[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

Let me know if I have made a mistake. Thank you again

**Comment by Astromis** [comment]
@Sahandfer 

You don't need to calculate argmax. The argmax shows you whether the token is, you know, empathic or not, as you got. The score you are looking for is a value of dimension that you got after applying argmax. 

By the way, it seems really weird that almost all tokens in a sequence are empathic. The reason might be that argmax doesn't care about values itself. For example, let for some token the probability score be [0.45, 0.55]. So argmax output will be 1 (empathic), but it can be clearly see, that score for its "empathiness" is very low. To get more realistic picture, try to use the threshold value instead of argmax, I mean if the value of second dimension doesn't exceed the threshold, say 90, than, the token is not empathic. In numpy setting you can do this as next `np.where(array[:,1] > 0.9)[0]` and you get all indices. Don't know unfortunately how to do equivalent in the torch. 

However, what you are looking for is those probability scores. You can measure "empathiness" like entropy, or just how maximum from these two values is far from 0.5.

**Comment by Sahandfer** [comment]
@Astromis 
I'm afraid that's might not be the case since the values are rather close to one another. For instance, for the response "Sorry to hear that. That must be really hard. Were you two close? I can't imagine losing my dad.", the logits after softmax are:
[[[0.0167, 0.0162],
         [0.0152, 0.0172],
         [0.0145, 0.0172],
         [0.0139, 0.0164],
         [0.0140, 0.0163],
         [0.0107, 0.0171],
         [0.0140, 0.0168],
         [0.0138, 0.0172],
         [0.0138, 0.0168],
         [0.0142, 0.0164],
         [0.0140, 0.0170],
         [0.0144, 0.0150],
         [0.0145, 0.0126],
         [0.0129, 0.0141],
         [0.0137, 0.0109],
         [0.0151, 0.0121],
         [0.0140, 0.0135],
         [0.0150, 0.0133],
         [0.0153, 0.0125],
         [0.0157, 0.0130],
         [0.0147, 0.0125],
         [0.0132, 0.0124],
         [0.0137, 0.0111],
         [0.0134, 0.0122],
         [0.0107, 0.0161],
         [0.0155, 0.0151],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163],
         [0.0167, 0.0163]]]
So it might be reasonable to choose  0.0165 or a similar value for the threshold but this threshold would work for a few cases and is not a general solution. However, I do believe that the progress so far has been enough to sort of find the rationales so thank you for your help. I will be closing the issue now. 

P.S. you can just change np to torch for the PyTorch equivalent :+1:

**Comment by ajyl** [comment]
Hi,

@Astromis @Sahandfer I just wanted to point out a couple of my observations.
First off, at least for calculating IOU-F1 scores, it seems like the author is applying argmax on to the logits directly: https://github.com/behavioral-data/Empathy-Mental-Health/blob/master/src/evaluation_utils.py#L76

As for the reason why most of the input sequence is appearing as empathetic ("Sorry to hear that. That must be really hard. Were you two close? I can't imagine losing my dad.", the output rationale is as follows:
[[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), I think it has to do with how the training data is being procssed: https://github.com/behavioral-data/Empathy-Mental-Health/blob/master/src/process_data.py#L63-L65

If I understand the above snippet correctly, it is marking the entire sequence as empathetic if there are no rationales. I'm not sure I understand the reason for this, and raised a question [here](https://github.com/behavioral-data/Empathy-Mental-Health/issues/11). But perhaps the behavior you are seeing is because of this? If you guys have any insight into this, it would be tremendously helpful.
